File: employee-service-scala/src/main/scala/com/example/employeeservice/repository/EmployeeRepository.scala
package com.example.employeeservice.repository

import com.example.employeeservice.model.Employee
import com.typesafe.scalalogging.LazyLogging
import scala.collection.concurrent.TrieMap
import scala.concurrent.{ExecutionContext, Future}

class EmployeeRepository(implicit ec: ExecutionContext) extends LazyLogging {
  // In-memory storage using TrieMap for thread safety
  private val employees = new TrieMap[String, Employee]()

  def findAll(): Future[Seq[Employee]] = {
    logger.debug(s"Finding all employees, total count: ${employees.size}")
    Future.successful(employees.values.toSeq)
  }

  def findById(id: String): Future[Option[Employee]] = {
    logger.debug(s"Finding employee by ID: $id")
    Future.successful(employees.get(id))
  }

  def save(employee: Employee): Future[Employee] = {
    logger.debug(s"Saving employee: $employee")
    employees.put(employee.id, employee)
    Future.successful(employee)
  }

  def update(id: String, employee: Employee): Future[Option[Employee]] = {
    logger.debug(s"Updating employee with ID: $id")
    if (employees.contains(id)) {
      val updatedEmployee = employee.copy(id = id)
      employees.put(id, updatedEmployee)
      Future.successful(Some(updatedEmployee))
    } else {
      logger.warn(s"Attempted to update non-existent employee with ID: $id")
      Future.successful(None)
    }
  }

  def delete(id: String): Future[Boolean] = {
    logger.debug(s"Deleting employee with ID: $id")
    Future.successful(employees.remove(id).isDefined)
  }
}

File: employee-service-ruby/app/models/employee_repository.rb
require_relative 'employee'
require 'logger'

class EmployeeRepository
  def initialize(logger)
    @employees = {}
    @logger = logger
  end
  
  def find_all
    @logger.info "Finding all employees, total count: #{@employees.size}"
    @employees.values
  end
  
  def find_by_id(id)
    @logger.info "Finding employee by ID: #{id}"
    @employees[id]
  end
  
  def save(employee)
    @logger.info "Saving employee: #{employee.to_h}"
    @employees[employee.id] = employee
    employee
  end
  
  def update(id, employee)
    @logger.info "Updating employee with ID: #{id}"
    if @employees.key?(id)
      employee.id = id
      @employees[id] = employee
      employee
    else
      @logger.warn "Attempted to update non-existent employee with ID: #{id}"
      nil
    end
  end
  
  def delete(id)
    @logger.info "Deleting employee with ID: #{id}"
    @employees.delete(id)
  end
  
  def add_sample_data
    @logger.info "Adding sample data to employee repository"
    
    sample_employees = [
      {
        first_name: "John",
        last_name: "Doe",
        email: "john.doe@example.com",
        phone: "555-123-4567",
        position: "Software Engineer",
        department: "Engineering",
        salary: 85000.00,
        hire_date: "2020-03-15"
      },
      {
        first_name: "Jane",
        last_name: "Smith",
        email: "jane.smith@example.com",
        phone: "555-987-6543",
        position: "Product Manager",
        department: "Product",
        salary: 95000.00,
        hire_date: "2019-07-10"
      },
      {
        first_name: "Michael",
        last_name: "Johnson",
        email: "michael.johnson@example.com",
        phone: "555-234-5678",
        position: "UX Designer",
        department: "Design",
        salary: 78000.00,
        hire_date: "2021-01-05"
      }
    ]
    
    sample_employees.each do |data|
      employee = Employee.new(data)
      save(employee)
    end
  end
end
